x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/FilterOperator.java

```diff
@@ -71,10 +71,18 @@ protected Page process(Page page) {
             positions = Arrays.copyOf(positions, rowCount);
 
             Block[] filteredBlocks = new Block[page.getBlockCount()];
-            for (int i = 0; i < page.getBlockCount(); i++) {
-                filteredBlocks[i] = page.getBlock(i).filter(positions);
+            boolean success = false;
+            try {
+                for (int i = 0; i < page.getBlockCount(); i++) {
+                    filteredBlocks[i] = page.getBlock(i).filter(positions);
+                }
+                success = true;
+            } finally {
+                Releasables.closeExpectNoException(page::releaseBlocks);
+                if (success == false) {
+                    Releasables.closeExpectNoException(filteredBlocks);
+                }
```

Comment:
Nice, this is much easier to reason about!

This should make the fix suggested in https://github.com/elastic/elasticsearch/pull/100278 obsolete. To confirm, you could remove `FilterOperatorTests::canLeak` and see if `FilterOperatorTests::testSimpleSmallInput` and ...::testSimpleLargeInput` still pass - or I can make a small follow-up PR that does that.

Reply:
Good point, I've disabled `allowLeaking` in both Limit and Filter operators in https://github.com/elastic/elasticsearch/pull/100465/commits/e1e7036cecb8d00e5b05ee66b346f8ce122b5bda